src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/RefDictionary.cs

```diff
@@ -0,0 +1,343 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Experimental.Collections
+{
+    public class RefDictionary<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey> // almost IReadOnlyDictionary<TKey, TValue>
+    {
+        private int[] _buckets;
+        private Entry[] _entries;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            public int next;
+        }
+
+        public RefDictionary(int capacity = 16)
+        {
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketIndex(TKey key) => (key.GetHashCode() & 0x7FFFFFFF) % _buckets.Length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(TKey key) => _buckets[GetBucketIndex(key)] - 1;
+
+        public int Count { get; private set; }
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public ref TValue this[TKey key]
+        {
+            get
+            {
+                Entry[] entries = _entries;
+                int entryIndex = GetEntryIndex(key);
+
+                while (entryIndex != -1)
+                {
+                    if (entries[entryIndex].key.Equals(key))
+                    {
+                        return ref entries[entryIndex].value;
+                    }
+                    entryIndex = entries[entryIndex].next;
+                }
+
+                if (Count == entries.Length)
+                {
+                    var count = Count;
+                    entries = new Entry[count * 2];
```

Comment:
I see this resize code in a couple spots; maybe worth making this a helper method instead

Reply:
Might also help perf to reduce the size of this method ...

Reply:
Oh, I think it's only one spot (the other code below is baseline implementation). Nevertheless my comment stands, similar to Dictionary.cs that has `Resize();`